cpp/src/tiledb/cxx_api/tiledb_cxx_array_util.cc

```diff
@@ -1,17 +1,195 @@
+    return -2;
+  }
+  catch (...) {
+    std::cout << "ArrayUtil::export_file_to_path, caught unknown error!" << std::endl;
+    return -3;
+  }
+
+  return 0;
+}
  
+int ArrayUtil::save_file_from_path(const std::string& file_uri, const std::string& input_path, const std::string& mime_type, const std::string& mime_encoding, const std::shared_ptr<tiledb::Context>& ctx) {
+  const std::string METADATA_SIZE_KEY = "file_size";
+  const std::string FILE_DIMENSION_NAME = "position";
+  const std::string FILE_ATTRIBUTE_NAME = "contents";
+  const std::string FILE_METADATA_MIME_TYPE_KEY = "mime";
+  const std::string FILE_METADATA_MIME_ENCODING_KEY = "mime_encoding";
+  const std::string METADATA_ORIGINAL_FILE_NAME = "original_file_name";
+  try {
+    if (input_path.empty()) {
+      std::cout << "ArrayUtil::save_file_from_path, input_path can not be empty!" << std::endl;
+      return -1;
+    }
+    std::shared_ptr<tiledb::VFS> vfs = std::shared_ptr<tiledb::VFS>(new tiledb::VFS(ctx));
+    tiledb::VFS::filebuf filebuf = tiledb::VFS::filebuf(vfs);
+    filebuf.open(input_path, std::ios::in);
+    std::istream os(&filebuf);
+    uint64_t nbytes = vfs->file_size(input_path);
+
+
+    uint64_t tile_extent = 1024;
+    if (nbytes > (1024ULL * 1024ULL * 1024ULL * 10ULL)) {
+      tile_extent = 1024ULL * 1024 * 100;
+    }
+    else if (nbytes > (1024UL * 1024 * 100)) {
+      tile_extent = 1024ULL * 1024 * 1;
+    }
+    else if (nbytes > (1024UL * 1024 * 1)) {
+      tile_extent = 1024ULL * 256;
+    }
+
+    std::shared_ptr<tiledb::ArraySchema> file_schema = std::shared_ptr<tiledb::ArraySchema>(new tiledb::ArraySchema(ctx, tiledb::ArrayType::TILEDB_DENSE));
+    tiledb::Domain dom(ctx);
+    std::array<uint64_t, 2> dim_domain = { 0,std::numeric_limits<uint64_t>::max() - tile_extent -2 };
+    tiledb::Dimension dim = tiledb::Dimension::create(ctx, FILE_DIMENSION_NAME, dim_domain, tile_extent);
+    dom.add_dimension(dim);
+    file_schema->set_domain(dom);
+    tiledb::Attribute attr = tiledb::Attribute::create_attribute(ctx, "contents", tiledb::DataType::TILEDB_UINT8);
+    attr.set_cell_val_num(1);
+    file_schema->add_attribute(attr);
+
+    tiledb::Array::create(file_uri, file_schema);
+
+    std::shared_ptr<Array> array = std::shared_ptr<Array>(new Array(ctx, file_uri, tiledb::QueryType::TILEDB_WRITE));
+    Query query(ctx, array, tiledb::QueryType::TILEDB_WRITE);
+
+    std::vector<uint8_t> contents;
+    contents.resize(nbytes);
+    std::array<uint64_t, 2> subarray = { 0,nbytes - 1 };
+    query.set_layout(tiledb::LayoutType::TILEDB_ROW_MAJOR);
+    query.set_subarray(subarray);
+    os.read((char*)contents.data(), nbytes);
+      
+    query.set_buffer(FILE_ATTRIBUTE_NAME, contents);
+    tiledb::QueryStatus query_status = query.submit();
+    if (query_status == tiledb::QueryStatus::TILEDB_COMPLETED) {
+      query.finalize();
+    }
+    else {
+      std::cout << "ArrayUtil::save_file_from_path, query not completed!" << std::endl;
```

Comment:
Shouldn't this be an error? We don't want them to think a file is written correctly if it is not actually complete... cc @Shelnutt2

Reply:
Agreed but this API is only for testing, we do not expect end users to call the write functionality yet only read from C#, so we have flexibility.

Reply:
added return statement when the write query is not complete

Reply:
Thanks